packages/flutter/lib/src/rendering/flex.dart

```diff
@@ -505,93 +591,47 @@ class RenderFlex extends RenderBox with ContainerRenderObjectMixin<RenderBox, Fl
-      // children, after the flexible children are fit into the available space,
-      // with the children sized using their max intrinsic dimensions.
-
-      // Get inflexible space using the max intrinsic dimensions of fixed children in the main direction.
-      final double availableMainSpace = extent;
-      int totalFlex = 0;
-      double inflexibleSpace = 0.0;
-      double maxCrossSize = 0.0;
-      RenderBox? child = firstChild;
-      while (child != null) {
-        final int flex = _getFlex(child);
-        totalFlex += flex;
-        late final double mainSize;
-        late final double crossSize;
-        if (flex == 0) {
-          switch (_direction) {
-            case Axis.horizontal:
-              mainSize = child.getMaxIntrinsicWidth(double.infinity);
-              crossSize = childSize(child, mainSize);
-            case Axis.vertical:
-              mainSize = child.getMaxIntrinsicHeight(double.infinity);
-              crossSize = childSize(child, mainSize);
-          }
-          inflexibleSpace += mainSize;
-          maxCrossSize = math.max(maxCrossSize, crossSize);
+      Size computeIntrinsicCrossExtentForChild(RenderBox child, BoxConstraints constraints) {
+        final bool isFlexChild = _getFlex(child) > 0;
+        switch (_direction) {
+          case Axis.horizontal:
+            final double width = isFlexChild
+              ? math.min(child.getMaxIntrinsicWidth(constraints.maxHeight), constraints.maxWidth)
+              : child.getMaxIntrinsicWidth(constraints.maxHeight);
+            return Size(width, childSize(child, width));
+          case Axis.vertical:
+            final double height = isFlexChild
+              ? math.min(child.getMaxIntrinsicHeight(constraints.maxWidth), constraints.maxHeight)
+              : child.getMaxIntrinsicHeight(constraints.maxWidth);
+            return Size(childSize(child, height), height);
         }
-        final FlexParentData childParentData = child.parentData! as FlexParentData;
-        child = childParentData.nextSibling;
       }
-
-      // Determine the spacePerFlex by allocating the remaining available space.
-      // When you're overconstrained spacePerFlex can be negative.
-      final double spacePerFlex = math.max(0.0, (availableMainSpace - inflexibleSpace) / totalFlex);
-
-      // Size remaining (flexible) items, find the maximum cross size.
-      child = firstChild;
-      while (child != null) {
-        final int flex = _getFlex(child);
-        if (flex > 0) {
-          maxCrossSize = math.max(maxCrossSize, childSize(child, spacePerFlex * flex));
-        }
-        final FlexParentData childParentData = child.parentData! as FlexParentData;
-        child = childParentData.nextSibling;
-      }
-
-      return maxCrossSize;
+      return _computeSizes(constraints: extentConstraints, layoutChild: computeIntrinsicCrossExtentForChild, computeBaseline: ChildLayoutHelper.getDryBaseline).size.crossAxisExtent;
```

Comment:
here and elsewhere in this method: these long method calls should probably be formatted over multiple lines.